File: core/c_ref/rnd.c
/*
 * Phase 1 (TASK-008.04) differential-test pilot.
 *
 * main.c's rnd() extracted verbatim, compiled here as the renamed-C-reference
 * side of the difftest harness (core/build.zig's compileRenamedCRef renames
 * this to c_rnd) so core/rnd_difftest.zig has something concrete to diff the
 * trivial core/rnd.zig passthrough against before any real TASK-011.* module
 * exists. Keep this in sync with main.c's rnd() by hand; when TASK-011 ports
 * the real RNG call sites, this pilot is retired in favor of the real module.
 */
#include <stdlib.h>

unsigned short rnd(unsigned short max)
{
#if (RAND_MAX < 0x7fff)
#error "rand returns too small values"
#elif (RAND_MAX == 0x7fff)
	return (unsigned short)((rand()*2) % (int)max);
#else
	return (unsigned short)(rand() % (int)max);
#endif
}

File: core/rnd.zig
const c = @cImport({
    @cInclude("stdlib.h");
});

/// Phase 1 (TASK-008.04) differential-test pilot: a trivial passthrough
/// "port" of main.c's rnd() (minus the headless-only rnd_call_count
/// bookkeeping added in TASK-008.02, which isn't part of the RNG formula).
/// Exists only to prove the renamed-C-reference difftest harness works
/// end-to-end before any real TASK-011.* module exists to diff for real.
pub fn rnd(max: u16) u16 {
    if (c.RAND_MAX < 0x7fff) @compileError("rand returns too small values");
    if (c.RAND_MAX == 0x7fff) {
        return @intCast(@mod(c.rand() *% 2, @as(c_int, max)));
    }
    return @intCast(@mod(c.rand(), @as(c_int, max)));
}
